File: Threads, Handlers, Message Queue and Loopers/We_Part/app/src/main/java/com/example/we_part/MainActivity.java
package com.example.we_part;

import androidx.annotation.WorkerThread;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private Button button1, button2;
    private workerThread workerThread;
    private String TAG = this.getClass().getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initviews();
        workerThread = new workerThread();
        workerThread.start();
        Log.d(TAG, "run");

    }

    private void initviews() {
        button1 = findViewById(R.id.button);
        button2 = findViewById(R.id.button2);

        button1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                workerThread.addTasktoMessagequeue(task1);
                Toast.makeText(MainActivity.this,"task 1" + Thread.currentThread().getName(),Toast.LENGTH_SHORT).show();
            }
        });

        button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                workerThread.addTasktoMessagequeue(task2);
                Toast.makeText(MainActivity.this,"task 2 " + Thread.currentThread().getName(),Toast.LENGTH_SHORT).show();

            }
        });
    }

    private Runnable task1 = new Runnable() {
        @Override
        public void run() {
            Log.d(TAG, "task 1" + Thread.currentThread().getName());
            try {
                Thread.sleep(3000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    };

    private Runnable task2 = new Runnable() {
        @Override
        public void run() {
            Log.d(TAG, "task 2 " + Thread.currentThread().getName());
            try {
                Thread.sleep(3000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    };
}

File: CodingTask/Pizza/app/src/main/java/com/example/pizzaandroid/ui/MainActivity.kt
package com.example.pizzaandroid.ui

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.pizzaandroid.DataModel.CartData
import com.example.pizzaandroid.R
import com.example.pizzaandroid.adaptor.CrustAdaptor
import com.example.pizzaandroid.adaptor.PizzaAdaptor
import com.example.pizzaandroid.data.PizzaRepo
import com.example.pizzaandroid.data.PizzaViewModel
import com.example.pizzaandroid.databinding.*
import com.example.pizzaandroid.localdatabase.Dao
import com.example.pizzaandroid.localdatabase.PizzaRoomDataBase
import com.example.pizzaandroid.localdatabase.PizzaViewModelFactory
import com.example.pizzaandroid.onCLiclListener.OnClickListener
import com.example.pizzaandroid.response.Crust
import com.example.pizzaandroid.response.Size
import com.google.android.material.bottomsheet.BottomSheetDialog

class MainActivity : AppCompatActivity(), OnClickListener {

    lateinit var activityMainBinding: ActivityMainBinding
    private lateinit var pizzaViewModel: PizzaViewModel
    private var list = mutableListOf<Crust>()
    lateinit var pizzaAdaptor: PizzaAdaptor
    lateinit var crustAdaptor: CrustAdaptor
    lateinit var bottomSheetActivityBinding: BottomSheetActivityBinding
    lateinit var pizzaRoomDataBase: PizzaRoomDataBase
    lateinit var crustItemLayoutBinding: CrustItemLayoutBinding
    lateinit var dao: Dao
    var count: Int = 0
    var sum: Int = 0
    var newData = CartData("",0.0,"",1)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)


        activityMainBinding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(activityMainBinding.root)


        pizzaRoomDataBase = PizzaRoomDataBase.getDataBaseObject(this)
        dao = pizzaRoomDataBase.getTaskDAO()
        val repo = PizzaRepo(dao)
        val viewModelFactory = PizzaViewModelFactory(repo)
        pizzaViewModel = ViewModelProviders.of(this, viewModelFactory).get(PizzaViewModel::class.java)


        activityMainBinding.recyclerView.apply {
            pizzaAdaptor = PizzaAdaptor(list, this@MainActivity)
            val linearLayoutManager = LinearLayoutManager(this@MainActivity)
            adapter = pizzaAdaptor
            layoutManager = linearLayoutManager
        }

        pizzaViewModel = ViewModelProvider(this).get(PizzaViewModel::class.java)
        pizzaViewModel.callApi()
        pizzaViewModel.liveData.observe(this, {

            list.clear()
            list.addAll(it)
            pizzaAdaptor.notifyDataSetChanged()
        })


    }

    override fun onClick(crust: Crust) {

        bottomSheetActivityBinding = BottomSheetActivityBinding.inflate(layoutInflater)

        val bottomDialog = BottomSheetDialog(
            this@MainActivity, R.style.BottomSheetDialogTheme
        )
        val bottomSheetView = LayoutInflater.from(applicationContext).inflate(
            R.layout.bottom_sheet_activity, bottomSheetActivityBinding.bottomsheet
//            findViewById<LinearLayout>(R.id.bottomsheet) ( Since we are using data binding we should not use find view by id )
        )
        var listonclick = arrayListOf<Size>()
        listonclick.addAll(crust.sizes)

//        setContentView(bottomSheetActivityBinding.root) ( bec of root error )

        bottomSheetActivityBinding.crustrecyclerview.apply {
            adapter = CrustAdaptor(listonclick, this@MainActivity)
            layoutManager = LinearLayoutManager(this@MainActivity)
        }


        bottomSheetActivityBinding.showcart.setOnClickListener {
            val intent = Intent(this@MainActivity, CartActivity::class.java)
            startActivity(intent)
        }

        newData.PizzaName =crust.name
        bottomDialog.setContentView(bottomSheetView)
        bottomDialog.show()
    }

    override fun onAddClick(size: Size) {

        count++
        bottomSheetActivityBinding.quntaity.text = newData.PizzaQunatity.toString()

        sum += size.price.toInt()

        bottomSheetActivityBinding.totalSum.text = "total amount " + sum.toString()
        if (count > 0) {
            bottomSheetActivityBinding.buttonshow.visibility = View.VISIBLE
        } else {
            bottomSheetActivityBinding.buttonshow.visibility = View.INVISIBLE

        }
        Toast.makeText(this, "the item is added to cart ", Toast.LENGTH_SHORT).show()
        newData.PizzaSize = size.name
        newData.PizzaPrice = size.price
        pizzaViewModel.addToDb(newData)
    }

    override fun onRemoveCLick(size: Size) {

        count--

        sum -= size.price.toInt()

        bottomSheetActivityBinding.quntaity.text = count.toString()
        bottomSheetActivityBinding.totalSum.text = "total amount " + sum.toString()
        Toast.makeText(this, "the item is removed from cart ", Toast.LENGTH_SHORT).show()

    }

}
